implementing snmp_notification_receiver for XR

tphoney opened the pull request.

Commit:
implementing snmp_notification_receiver for XR

tphoney commented:
@chburket another one for you to look over :)

mikewiebe commented on lib/cisco_node_utils/cmd_ref/snmp_notification_receiver.yaml:
```diff
@@ -1,52 +1,48 @@
 
 source_interface:
-  get_command: "show running-config snmp all"
+  _exclude: [ios_xr]
```
Why are you excluding this on XR?  From a quick reading of the docs, it appears that this is supported for XR.
- XR Reference: http://www.cisco.com/en/US/docs/ios_xr_sw/iosxr_r3.7/system_management/command/reference/yr37snmp.html#wp1151429
  - Search for `snmp-server trap-source`
- Compare this to Nexus: http://www.cisco.com/c/en/us/td/docs/switches/datacenter/sw/4_2/nx-os/system_management/configuration/guide/sm_nx_os_cli/sm_9snmp.html#wp1088298

mikewiebe commented on lib/cisco_node_utils/cmd_ref/snmp_notification_receiver.yaml:
```diff
@@ -1,52 +1,48 @@
 
 vrf:
-  get_command: "show running-config snmp all"
+  _exclude: [ios_xr]
```
Again, why exclude on XR?

http://www.cisco.com/en/US/docs/ios_xr_sw/iosxr_r3.7/system_management/command/reference/yr37snmp.html#wp1151429
- Search for `snmp-server vrf`

tphoney commented:
@mikewiebe i was unaware that the CLI was all over the place, previously this was all one CLI command, i will fix these omissions. apologies for the ping pong.

tphoney commented:
@mikewiebe @chrisvanheuveln i have addressed the 2 missing parameters. I have found one thing of note. On the EXR you can only set one snmp-server trap-source.
cisco-network-node-utils git:(snmp_notification_reciever) ✗ bundle exec ruby tests/test_snmp_notification_receiver.rb -v -- j2nvya0j6ll9jel.delivery.puppetlabs.net:57777 root lab
Run options: -v -- --seed 64322

# Running:

Node under test:
- name  - j2nvya0j6ll9jel
- type  - R-IOSXRV9000-CH
- image - 6.0.0.20I

TestSnmpNotificationReceiver#test_create_destroy_multiple = 118.72 s = .
TestSnmpNotificationReceiver#test_create_destroy_single = 69.11 s = .
TestSnmpNotificationReceiver#test_create_destroy_single_ipv6 = 46.36 s = .

Finished in 243.308521s, 0.0123 runs/s, 0.2261 assertions/s.

3 runs, 55 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest to /Users/tp/workspace/git/EXR/cisco_devenv/cisco-network-node-utils/coverage. 2357 / 4474 LOC (52.68%) covered.

mikewiebe commented on lib/cisco_node_utils/snmp_notification_receiver.rb:
```diff
@@ -146,11 +154,22 @@ def security
 
     def source_interface
       val = config_get('snmp_notification_receiver', 'source_interface', @name)
+
```
Unnecessary whitespace?

mikewiebe commented on lib/cisco_node_utils/snmp_notification_receiver.rb:
```diff
@@ -146,11 +154,22 @@ def security
+        # get all vrf
+        all_vrf = config_get('snmp_notification_receiver', 'vrf_all')
+        all_vrf.each do |vrf|
+          bla = config_get('snmp_notification_receiver', 'vrf_values', vrf: vrf)
```
Can we use something a bit more descriptive then `bla` :).  What kind of data is being gathered here?

chburket commented on lib/cisco_node_utils/snmp_notification_receiver.rb:
```diff
@@ -21,7 +21,7 @@ module Cisco
     attr_reader :name
 
-    def initialize(name,
+    def initialize(name, # rubocop:disable all
```
Why is the `rubocop:disable all` needed?

mikewiebe replied:
Good catch.. +1 on why this is needed?

mikewiebe commented:
Few nits but 👍 for the nxos impact.  @chburket can you take a look for XR?

chburket commented on tests/test_snmp_notification_receiver.rb:
```diff
@@ -100,7 +100,7 @@ def test_create_destroy_single_ipv6
     assert_equal('ab', Cisco::SnmpNotificationReceiver.receivers[id].username)
     assert_equal('3', Cisco::SnmpNotificationReceiver.receivers[id].version)
-    assert_equal('red', Cisco::SnmpNotificationReceiver.receivers[id].vrf)
+    assert_equal('red', Cisco::SnmpNotificationReceiver.receivers[id].vrf) # rubocop:disable Metrics/LineLength
```
This line is shorter than lines around it, was rubocop complaining about its length?

chburket commented on tests/test_snmp_notification_receiver.rb:
```diff
@@ -140,12 +140,12 @@ def test_create_destroy_multiple
     assert_equal('ab', Cisco::SnmpNotificationReceiver.receivers[id].username)
     assert_equal('3', Cisco::SnmpNotificationReceiver.receivers[id].version)
-    assert_equal('red', Cisco::SnmpNotificationReceiver.receivers[id].vrf)
+    assert_equal('red', Cisco::SnmpNotificationReceiver.receivers[id].vrf) # rubocop:disable Metrics/LineLength
```
same question here

chburket commented on tests/test_snmp_notification_receiver.rb:
```diff
@@ -157,7 +157,7 @@ def test_create_destroy_multiple
     assert_equal('cd', Cisco::SnmpNotificationReceiver.receivers[id2].username)
     assert_equal('3', Cisco::SnmpNotificationReceiver.receivers[id2].version)
-    assert_equal('red', Cisco::SnmpNotificationReceiver.receivers[id2].vrf)
+    assert_equal('red', Cisco::SnmpNotificationReceiver.receivers[id2].vrf) # rubocop:disable Metrics/LineLength
```
and here

chburket commented:
:+1:  except for a few rubocop questions

tphoney commented:
@mikewiebe @chburket 
I have made the following changes according to your recommendations:
- Unnecessary whitespace?  fixed
- Can we use something a bit more descriptive then bla :). What kind of data is being gathered here? fixed
- The reason for disabling rubocop on this: following errors that i dont think should be corrected
  lib/cisco_node_utils/snmp_notification_receiver.rb:24:5: C: Cyclomatic complexity for initialize is too high. [24/23]
  def initialize(name,
  ^^^
  lib/cisco_node_utils/snmp_notification_receiver.rb:24:5: C: Method has too many lines. [56/50]
  def initialize(name,
  ^^^
  lib/cisco_node_utils/snmp_notification_receiver.rb:24:5: C: Perceived complexity for initialize is too high. [26/24]
  def initialize(name,
  ^^^
- fixed the rubocop issues in tests/test_snmp_notification_receiver.rb
  
  cisco-network-node-utils git:(snmp_notification_reciever) ✗ bundle exec ruby tests/test_snmp_notification_receiver.rb -v -- xgop8ca555i662r.delivery.puppetlabs.net:57777 root lab
  Run options: -v -- --seed 48114

# Running:

Node under test:
- name  - xgop8ca555i662r
- type  - R-IOSXRV9000-CH
- image - 6.0.0.20I

TestSnmpNotificationReceiver#test_create_destroy_single_ipv6 = 60.80 s = .
TestSnmpNotificationReceiver#test_create_destroy_single = 48.20 s = .
TestSnmpNotificationReceiver#test_create_destroy_multiple = 89.24 s = .

Finished in 205.909999s, 0.0146 runs/s, 0.2671 assertions/s.

3 runs, 55 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest to /Users/tp/workspace/git/EXR/cisco_devenv/cisco-network-node-utils/coverage. 2476 / 4995 LOC (49.57%) covered.

chburket commented:
I'm ok with everything except the `# rubocop:disable all`.  At the very least, I'd rather see each issue disabled separately, so we don't allow every error possible in this method from now on.  But to be honest, I'd rather the errors just be fixed.  But if @mikewiebe disagrees, I'll be ok with that, too.

mikewiebe commented:
I agree with @chburket  that we should not just disable all cops, but only disable the specific cop for the affected lines of code.

tphoney commented:
I have disabled the pertinent rubocop failures and re-pushed. Apologies for the churn. @mikewiebe @chburket

chburket commented:
:+1:

mikewiebe commented:
👍 Verified on n3k(I2,I3), n5k, n6k, n7k, n8k, n9k(I2, I4)
